### ISSUE TITLE
Adding an implied 'await sleep(0)' after scope.cancel()

### ISSUE THREAD

**Issue by arthurxlee** [agronholm/anyio#166]
If I would `anyio.run()` the following:
```python
async def fn1():
    async with open_cancel_scope(shield=True) as scope:
        print(f'fn1: scope: cancelling scope: {scope.cancel_called=}')
        await scope.cancel()

        # await sleep(0)
        print(f'fn1: scope: scope cancelled: {scope.cancel_called=} (!!)')

async def fn2():
    async with open_cancel_scope() as scope1:
        async with open_cancel_scope(shield=True) as scope2:
            print(f'fn2: scope2: cancelling scope1: {scope1.cancel_called=}')
            await scope1.cancel()
            await sleep(0)
            print(f'fn2: scope2: shielded from scope1: {scope1.cancel_called=}')

        # await sleep(0)
        print(f'fn2: scope1: scope1 cancelled: {scope1.cancel_called=} (!!)')
```
I will receive the (surprising) output:
```
fn1: scope: cancelling scope: scope.cancel_called=False
fn1: scope: scope cancelled: scope.cancel_called=True (!!)

fn2: scope2: cancelling scope1: scope1.cancel_called=False
fn2: scope2: shielded from scope1: scope1.cancel_called=True
fn2: scope1: scope1 cancelled: scope1.cancel_called=True (!!)
```

But my expectation is:
```
fn1: scope: cancelling scope: scope.cancel_called=False

fn2: scope2: cancelling scope1: scope1.cancel_called=False
fn2: scope2: shielded from scope1: scope1.cancel_called=True
```

To meet my expectation, I would have to insert `await sleep(0)` in the positions I had indicated above.  I am not sure if it is even possible, but I would like to ask `anyio` to implicitly include the `await sleep(0)` (or equivalent) instead.

**Comment by agronholm** [comment]
Sorry for taking this long to respond, but I don't think AnyIO is doing anything wrong here. Here's the corresponding trio version:
```python3
from trio import CancelScope, run, sleep


async def fn1():
    with CancelScope(shield=True) as scope:
        print(f'fn1: scope: cancelling scope: {scope.cancel_called}')
        scope.cancel()
        print(f'fn1: scope: scope cancelled: {scope.cancel_called} (!!)')


async def fn2():
    with CancelScope() as scope1:
        with CancelScope(shield=True) as scope2:
            print(f'fn2: scope2: cancelling scope1: {scope1.cancel_called}')
            scope1.cancel()
            await sleep(0)
            print(f'fn2: scope2: shielded from scope1: {scope1.cancel_called}')

        print(f'fn2: scope1: scope1 cancelled: {scope1.cancel_called} (!!)')

run(fn1)
run(fn2)
```
And its output:
```python3
fn1: scope: cancelling scope: False
fn1: scope: scope cancelled: True (!!)
fn2: scope2: cancelling scope1: False
fn2: scope2: shielded from scope1: True
fn2: scope1: scope1 cancelled: True (!!)
```

**Comment by agronholm** [comment]
I'd like to add that `CancelScope.cancel()` is a coroutine function only because of curio, and curio support is being dropped. I can understand that you expect the `cancel()` to be a switch point, but making it that would surely break somebody else's expectations at this point. This will all be fixed in v3.0 where many former coroutine functions (including `CancelScope.cancel()`) will be turned into regular functions.

**Comment by arthurxlee** [comment]
Thank you for your response.  Given the roadmap towards v3.0, I do agree with you that this will stop being be a relevant issue.  Looking forward to the new version.
